src/main/scala/scalafix/internal/sbt/ScalafixInterface.scala

```diff
@@ -112,31 +122,42 @@ class ScalafixInterface private (
 }
 
 object ScalafixInterface {
-  private class LazyValue[T](thunk: () => T) extends (() => T) {
-    private lazy val _value = scala.util.Try(thunk())
-    override def apply(): T = _value.get
-  }
-  private val fromToolClasspathMemo: BlockingCache[
-    (String, Seq[ModuleID], Seq[Repository]),
-    ScalafixInterface
-  ] = new BlockingCache
+
+  type Cache = BlockingCache[
+    (
+        String, // scalafixScalaBinaryVersion
+        Option[Arg.ToolClasspath]
+    ),
+    (
+        ScalafixInterface,
+        Seq[Long] // mutable files stamping
+    )
+  ]
 
   private[scalafix] val defaultLogger: Logger = ConsoleLogger(System.out)
 
-  def fromToolClasspath(
+  def provider(
+      cache: Cache,
       scalafixScalaBinaryVersion: String,
-      scalafixDependencies: Seq[ModuleID],
-      scalafixCustomResolvers: Seq[Repository],
+      toolClasspath: Arg.ToolClasspath,
       logger: Logger,
       callback: ScalafixMainCallback
-  ): () => ScalafixInterface =
-    new LazyValue({ () =>
-      fromToolClasspathMemo.getOrElseUpdate(
-        (
-          scalafixScalaBinaryVersion,
-          scalafixDependencies,
-          scalafixCustomResolvers
-        ), {
+  ): () => ScalafixInterface = { () =>
+
+    // Build or retrieve from the cache the scalafix interface that can run
+    // built-in rules. This is the most costly instanciation, which should be
+    // shared as much as possible.
+    val (buildinRulesInterface, _) = cache.compute(
+      (
+        scalafixScalaBinaryVersion,
+        None
```

Comment:
Memoizing first an interface per Scala version (ie.e ignoring the requested tool classpath) is key for the tool classpath instances created below to share the common classloader with the entire Scala standard library.